### DISCUSSION TITLE
`describeServer` field for reporting software version(s)?

### DISCUSSION THREAD

**Discussion by DavidBuchanan314** [bluesky-social/atproto#1249] [Protocol (atproto)]
A free-form text field would be enough I suppose (like with the HTTP `Server` header). It would be an informational field for developers to see, useful for diagnosing federation issues in the wild, and also generally nice for collecting stats (e.g. https://atscan.net/)

**Comment by snarfed** [comment]
Yes! Also relevant: https://nodeinfo.diaspora.software/

**Reply by DavidBuchanan314** [reply]
I notice you serve a nodeinfo at https://atproto.brid.gy/.well-known/nodeinfo - I might do something similar for millipds

**Reply by DavidBuchanan314** [reply]
also @ngerakines just showed me https://github.com/jhass/nodeinfo/pull/100

**Reply by snarfed** [reply]
Hah, yes! Would help set a precedent for nodeinfo to expand beyond AP/fediverse-centric.

**Comment by DavidBuchanan314** [comment]
A related question: https://github.com/bluesky-social/atproto/discussions/1139

**Comment by bnewbold** [comment]
Better structured metadata would be good! Could be `describeServer`, could be a new `com.atproto.*` endpoint (eg, to be more generic to multiple network roles; `describeServer` is a bit PDS-centric); HTTP `Server` and HTML metadata would be good regardless of more granular solutions.

FWIW, the current typescript PDS implementation does stuff a git commit in the healthcheck endpoint (eg, https://bsky.social/xrpc/_health). But we shouldn't lean on that too much, don't want it to become a de-facto standard.

**Comment by de-cache** [comment]
I've sketched out a schema for an expanded describeServer [here](https://tangled.org/strings/hexmani.ac/3mdjtphd55t22), which mainly adds more metadata and support for relays. Its new fields are inspired by the Mastodon API's instance info endpoint and NodeInfo (which, in my opinion, should be adopted in tandem with describeServer).

I drew this up partly because my [PDS version-tracking bot](https://bsky.app/profile/did:plc:wo6c3x2ghqnmlcokhubufytk) identifies PDS implementations using the versioning scheme outputted by the given instance, which simply will not scale as more third-party implementations (either PDSes or relays) become ready for production usage over time.

**Comment by verdverm** [comment]
When deploying Nginx (i.e.), a security recommendation is to not return the software version. How do you think about the security implications of returning the PDS version? Can this be optional?

**Reply by DavidBuchanan314** [reply]
IMHO obfuscating software versions is an exercise in security-through-obscurity and is rarely beneficial.

The main reason it's nice to know a software version in the first place is because it can affect observable behaviours. And if you can observe the behaviours then you can infer the software version (i.e. "fingerprinting").
